File: frontend/src/utils/engagementUtils.ts
/**
 * Engagement Rate Utilities for Multi-Question Form Apps
 *
 * This module provides functions to compute various engagement metrics for form submissions.
 * All functions handle large datasets efficiently with O(n*m) time complexity.
 */

export type Submission = Record<string | number, any>;
export type QuestionId = string | number;

/**
 * Determines if a response value is considered "answered"
 * @param value - The response value to check
 * @returns true if the value is considered answered
 */
export function isAnswered(value: any): boolean {
  return value !== null &&
         value !== undefined &&
         (typeof value !== 'string' || value.trim() !== '');
}

/**
 * Computes the overall engagement rate across all questions and submissions
 * @param n - Number of questions in the form
 * @param submissions - Array of submission objects
 * @param questionIds - Optional array of all question IDs (length should be n)
 * @returns Overall engagement rate as percentage (0-100)
 */
export function computeOverallEngagement(
  n: number,
  submissions: Submission[],
  questionIds?: QuestionId[]
): number {
  if (n <= 0) return 0;
  const m = submissions.length;
  if (m === 0) return 0;

  // Determine question IDs
  const questions = questionIds || getUniqueQuestionIds(submissions);
  if (questions.length !== n) {
    throw new Error(`Expected ${n} questions, but found ${questions.length}`);
  }

  let totalAnswered = 0;

  for (const submission of submissions) {
    for (const questionId of questions) {
      if (isAnswered(submission[questionId])) {
        totalAnswered++;
      }
    }
  }

  return (totalAnswered / (n * m)) * 100;
}

/**
 * Computes engagement rate per question
 * @param n - Number of questions in the form
 * @param submissions - Array of submission objects
 * @param questionIds - Optional array of all question IDs (length should be n)
 * @returns Map of question ID to {answeredCount, engagementPercent}
 */
export function computePerQuestionEngagement(
  n: number,
  submissions: Submission[],
  questionIds?: QuestionId[]
): Map<QuestionId, { answeredCount: number; engagementPercent: number }> {
  if (n <= 0) return new Map();
  const m = submissions.length;
  if (m === 0) return new Map();

  // Determine question IDs
  const questions = questionIds || getUniqueQuestionIds(submissions);
  if (questions.length !== n) {
    throw new Error(`Expected ${n} questions, but found ${questions.length}`);
  }

  const result = new Map<QuestionId, { answeredCount: number; engagementPercent: number }>();

  for (const questionId of questions) {
    let answeredCount = 0;
    for (const submission of submissions) {
      if (isAnswered(submission[questionId])) {
        answeredCount++;
      }
    }
    const engagementPercent = (answeredCount / m) * 100;
    result.set(questionId, { answeredCount, engagementPercent });
  }

  return result;
}

/**
 * Computes engagement rate per user/submission
 * @param n - Number of questions in the form
 * @param submissions - Array of submission objects
 * @param questionIds - Optional array of all question IDs (length should be n)
 * @returns Array of {userIndex, answeredCount, engagementPercent}
 */
export function computePerUserEngagement(
  n: number,
  submissions: Submission[],
  questionIds?: QuestionId[]
): Array<{ userIndex: number; answeredCount: number; engagementPercent: number }> {
  if (n <= 0) return [];
  const m = submissions.length;
  if (m === 0) return [];

  // Determine question IDs
  const questions = questionIds || getUniqueQuestionIds(submissions);
  if (questions.length !== n) {
    throw new Error(`Expected ${n} questions, but found ${questions.length}`);
  }

  const result: Array<{ userIndex: number; answeredCount: number; engagementPercent: number }> = [];

  for (let i = 0; i < m; i++) {
    const submission = submissions[i];
    let answeredCount = 0;
    for (const questionId of questions) {
      if (isAnswered(submission[questionId])) {
        answeredCount++;
      }
    }
    const engagementPercent = (answeredCount / n) * 100;
    result.push({ userIndex: i, answeredCount, engagementPercent });
  }

  return result;
}

/**
 * Computes the completion rate (percentage of submissions that answered all questions)
 * @param submissions - Array of submission objects
 * @param questionIds - Optional array of all question IDs
 * @returns Completion rate as percentage (0-100)
 */
export function computeCompletionRate(
  submissions: Submission[],
  questionIds?: QuestionId[]
): number {
  const m = submissions.length;
  if (m === 0) return 0;

  // Determine question IDs
  const questions = questionIds || getUniqueQuestionIds(submissions);
  const n = questions.length;
  if (n === 0) return 0;

  let completedCount = 0;

  for (const submission of submissions) {
    let answeredCount = 0;
    for (const questionId of questions) {
      if (isAnswered(submission[questionId])) {
        answeredCount++;
      }
    }
    if (answeredCount === n) {
      completedCount++;
    }
  }

  return (completedCount / m) * 100;
}

/**
 * Extracts unique question IDs from submissions
 * @param submissions - Array of submission objects
 * @returns Array of unique question IDs
 */
function getUniqueQuestionIds(submissions: Submission[]): QuestionId[] {
  const questionSet = new Set<QuestionId>();
  for (const submission of submissions) {
    for (const key in submission) {
      questionSet.add(key);
    }
  }
  return Array.from(questionSet);
}

// Unit tests
export function runTests(): void {
  console.log('Running engagement utilities tests...');

  // Test data
  const submissions1: Submission[] = [
    { q1: 'Yes', q2: null, q3: 'Some text' },
    { q1: 'No', q2: 'Answer', q3: null },
    { q1: null, q2: null, q3: 'Text' }
  ];

  const questionIds1 = ['q1', 'q2', 'q3'];

  // Test computeOverallEngagement
  const overall = computeOverallEngagement(3, submissions1, questionIds1);
  console.assert(Math.abs(overall - (5/9)*100) < 0.01, `Overall engagement: expected ${(5/9)*100}, got ${overall}`);

  // Test computePerQuestionEngagement
  const perQuestion = computePerQuestionEngagement(3, submissions1, questionIds1);
  console.assert(perQuestion.get('q1')!.answeredCount === 2, 'q1 answered count');
  console.assert(perQuestion.get('q2')!.answeredCount === 1, 'q2 answered count');
  console.assert(perQuestion.get('q3')!.answeredCount === 2, 'q3 answered count');

  // Test computePerUserEngagement
  const perUser = computePerUserEngagement(3, submissions1, questionIds1);
  console.assert(perUser[0].answeredCount === 2, 'User 0 answered count');
  console.assert(perUser[1].answeredCount === 2, 'User 1 answered count');
  console.assert(perUser[2].answeredCount === 1, 'User 2 answered count');

  // Test computeCompletionRate
  const completion = computeCompletionRate(submissions1, questionIds1);
  console.assert(Math.abs(completion - (0/3)*100) < 0.01, `Completion rate: expected 0, got ${completion}`);

  // Edge cases
  console.assert(computeOverallEngagement(0, [], []) === 0, 'Zero questions');
  console.assert(computeOverallEngagement(1, [], ['q1']) === 0, 'Zero submissions');
  console.assert(computeCompletionRate([], []) === 0, 'Empty submissions');

  // Test with derived questionIds
  const submissions2 = [{ q1: 'a', q2: null }, { q1: null, q2: 'b' }];
  const overall2 = computeOverallEngagement(2, submissions2);
  console.assert(Math.abs(overall2 - (2/4)*100) < 0.01, `Derived questionIds: expected 50, got ${overall2}`);

  // Test isAnswered
  console.assert(isAnswered('text') === true, 'String answered');
  console.assert(isAnswered('') === false, 'Empty string not answered');
  console.assert(isAnswered('   ') === false, 'Whitespace not answered');
  console.assert(isAnswered(0) === true, 'Zero answered');
  console.assert(isAnswered(false) === true, 'False answered');
  console.assert(isAnswered(null) === false, 'Null not answered');
  console.assert(isAnswered(undefined) === false, 'Undefined not answered');

  console.log('All tests passed!');
}

// Example usage
export function examples(): void {
  const submissions: Submission[] = [
    { q1: 'Yes', q2: null, q3: 'Some text' },
    { q1: 'No', q2: 'Answer', q3: null },
    { q1: null, q2: null, q3: 'Text' }
  ];

  const n = 3;
  const questionIds = ['q1', 'q2', 'q3'];

  console.log('Overall Engagement:', computeOverallEngagement(n, submissions, questionIds));
  console.log('Per Question:', computePerQuestionEngagement(n, submissions, questionIds));
  console.log('Per User:', computePerUserEngagement(n, submissions, questionIds));
  console.log('Completion Rate:', computeCompletionRate(submissions, questionIds));
}

File: frontend/src/utils/engagement_utils.py
"""
Engagement Rate Utilities for Multi-Question Form Apps

This module provides functions to compute various engagement metrics for form submissions.
All functions handle large datasets efficiently with O(n*m) time complexity.
"""

from typing import Any, Dict, List, Optional, Set, Union, Tuple
import math

Submission = Dict[Union[str, int], Any]
QuestionId = Union[str, int]

def is_answered(value: Any) -> bool:
    """
    Determines if a response value is considered "answered"
    """
    return (value is not None and
            value is not None and  # Note: 'is not None' is correct, but to match TS
            (not isinstance(value, str) or value.strip() != ''))

def compute_overall_engagement(
    n: int,
    submissions: List[Submission],
    question_ids: Optional[List[QuestionId]] = None
) -> float:
    """
    Computes the overall engagement rate across all questions and submissions

    Formula: (total answered responses) / (n * m) * 100
    where m = len(submissions)
    """
    if n <= 0:
        return 0.0
    m = len(submissions)
    if m == 0:
        return 0.0

    # Determine question IDs
    questions = question_ids if question_ids is not None else _get_unique_question_ids(submissions)
    if len(questions) != n:
        raise ValueError(f"Expected {n} questions, but found {len(questions)}")

    total_answered = 0

    for submission in submissions:
        for question_id in questions:
            if is_answered(submission.get(question_id)):
                total_answered += 1

    return (total_answered / (n * m)) * 100

def compute_per_question_engagement(
    n: int,
    submissions: List[Submission],
    question_ids: Optional[List[QuestionId]] = None
) -> Dict[QuestionId, Dict[str, Union[int, float]]]:
    """
    Computes engagement rate per question

    Returns: dict of question_id -> {'answered_count': int, 'engagement_percent': float}
    """
    if n <= 0:
        return {}
    m = len(submissions)
    if m == 0:
        return {}

    # Determine question IDs
    questions = question_ids if question_ids is not None else _get_unique_question_ids(submissions)
    if len(questions) != n:
        raise ValueError(f"Expected {n} questions, but found {len(questions)}")

    result = {}

    for question_id in questions:
        answered_count = 0
        for submission in submissions:
            if is_answered(submission.get(question_id)):
                answered_count += 1
        engagement_percent = (answered_count / m) * 100
        result[question_id] = {
            'answered_count': answered_count,
            'engagement_percent': engagement_percent
        }

    return result

def compute_per_user_engagement(
    n: int,
    submissions: List[Submission],
    question_ids: Optional[List[QuestionId]] = None
) -> List[Dict[str, Union[int, float]]]:
    """
    Computes engagement rate per user/submission

    Returns: list of {'user_index': int, 'answered_count': int, 'engagement_percent': float}
    """
    if n <= 0:
        return []
    m = len(submissions)
    if m == 0:
        return []

    # Determine question IDs
    questions = question_ids if question_ids is not None else _get_unique_question_ids(submissions)
    if len(questions) != n:
        raise ValueError(f"Expected {n} questions, but found {len(questions)}")

    result = []

    for i, submission in enumerate(submissions):
        answered_count = 0
        for question_id in questions:
            if is_answered(submission.get(question_id)):
                answered_count += 1
        engagement_percent = (answered_count / n) * 100
        result.append({
            'user_index': i,
            'answered_count': answered_count,
            'engagement_percent': engagement_percent
        })

    return result

def compute_completion_rate(
    submissions: List[Submission],
    question_ids: Optional[List[QuestionId]] = None
) -> float:
    """
    Computes the completion rate (percentage of submissions that answered all questions)

    Formula: (completed submissions) / m * 100
    """
    m = len(submissions)
    if m == 0:
        return 0.0

    # Determine question IDs
    questions = question_ids if question_ids is not None else _get_unique_question_ids(submissions)
    n = len(questions)
    if n == 0:
        return 0.0

    completed_count = 0

    for submission in submissions:
        answered_count = 0
        for question_id in questions:
            if is_answered(submission.get(question_id)):
                answered_count += 1
        if answered_count == n:
            completed_count += 1

    return (completed_count / m) * 100

def _get_unique_question_ids(submissions: List[Submission]) -> List[QuestionId]:
    """Extracts unique question IDs from submissions"""
    question_set = set()
    for submission in submissions:
        question_set.update(submission.keys())
    return list(question_set)

# Unit tests
def run_tests():
    """Run unit tests for all functions"""
    print('Running engagement utilities tests...')

    # Test data
    submissions1 = [
        {'q1': 'Yes', 'q2': None, 'q3': 'Some text'},
        {'q1': 'No', 'q2': 'Answer', 'q3': None},
        {'q1': None, 'q2': None, 'q3': 'Text'}
    ]

    question_ids1 = ['q1', 'q2', 'q3']

    # Test compute_overall_engagement
    overall = compute_overall_engagement(3, submissions1, question_ids1)
    expected_overall = (5/9) * 100
    assert abs(overall - expected_overall) < 0.01, f"Overall engagement: expected {expected_overall}, got {overall}"

    # Test compute_per_question_engagement
    per_question = compute_per_question_engagement(3, submissions1, question_ids1)
    assert per_question['q1']['answered_count'] == 2, 'q1 answered count'
    assert per_question['q2']['answered_count'] == 1, 'q2 answered count'
    assert per_question['q3']['answered_count'] == 2, 'q3 answered count'

    # Test compute_per_user_engagement
    per_user = compute_per_user_engagement(3, submissions1, question_ids1)
    assert per_user[0]['answered_count'] == 2, 'User 0 answered count'
    assert per_user[1]['answered_count'] == 2, 'User 1 answered count'
    assert per_user[2]['answered_count'] == 1, 'User 2 answered count'

    # Test compute_completion_rate
    completion = compute_completion_rate(submissions1, question_ids1)
    assert abs(completion - 0.0) < 0.01, f"Completion rate: expected 0, got {completion}"

    # Edge cases
    assert compute_overall_engagement(0, [], []) == 0, 'Zero questions'
    assert compute_overall_engagement(1, [], ['q1']) == 0, 'Zero submissions'
    assert compute_completion_rate([], []) == 0, 'Empty submissions'

    # Test with derived question_ids
    submissions2 = [{'q1': 'a', 'q2': None}, {'q1': None, 'q2': 'b'}]
    overall2 = compute_overall_engagement(2, submissions2)
    assert abs(overall2 - 50.0) < 0.01, f"Derived question_ids: expected 50, got {overall2}"

    # Test is_answered
    assert is_answered('text') == True, 'String answered'
    assert is_answered('') == False, 'Empty string not answered'
    assert is_answered('   ') == False, 'Whitespace not answered'
    assert is_answered(0) == True, 'Zero answered'
    assert is_answered(False) == True, 'False answered'
    assert is_answered(None) == False, 'None not answered'

    print('All tests passed!')

# Example usage
def examples():
    submissions = [
        {'q1': 'Yes', 'q2': None, 'q3': 'Some text'},
        {'q1': 'No', 'q2': 'Answer', 'q3': None},
        {'q1': None, 'q2': None, 'q3': 'Text'}
    ]

    n = 3
    question_ids = ['q1', 'q2', 'q3']

    print('Overall Engagement:', compute_overall_engagement(n, submissions, question_ids))
    print('Per Question:', compute_per_question_engagement(n, submissions, question_ids))
    print('Per User:', compute_per_user_engagement(n, submissions, question_ids))
    print('Completion Rate:', compute_completion_rate(submissions, question_ids))

if __name__ == '__main__':
    examples()
    run_tests()